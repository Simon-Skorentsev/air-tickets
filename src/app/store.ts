import { configureStore, ThunkAction, Action } from '@reduxjs/toolkit';
import ticketsListReduser from '../features/ticketsList/ticketsListSlice';

export const store = configureStore({
  reducer: {
    ticketsList: ticketsListReduser,
  },
});

export type Store = typeof store;
export type AppDispatch = typeof store.dispatch;
export type RootState = ReturnType<typeof store.getState>;
export type AppThunk<ReturnType = void> = ThunkAction<
  ReturnType,
  RootState,
  unknown,
  Action<string>
>;
