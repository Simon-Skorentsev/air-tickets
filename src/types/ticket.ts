export interface Ticket {
    origin: string;
    origin_name: string;
    destination: string;
    destination_name: string;
    departure_date: string;
    departure_time: string;
    arrival_date: string;
    arrival_time: string;
    carrier: string;
    stops: number;
    price: number;
    initialPrice: number;
};

export type ExchangeRates = 1| 73 | 84;
export type Currency = "rub" | "usd" | "eur";
export type CountOfTransfer = number | "all";

export interface TicketsState {
    tickets: Ticket[];
    filteredTickets: Ticket[];
    loading: boolean;
    error: null | string;
};