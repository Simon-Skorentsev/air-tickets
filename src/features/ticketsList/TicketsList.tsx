import React, { useEffect } from 'react';
import { fetchTickets } from '../../API/ticketsListAPI';
import { useAppDispatch, useAppSelector } from '../../app/hooks';
import styles from "./TicketsList.module.css";

const TicketsList: React.FC = () => {
  const dispatch = useAppDispatch();
  const {error, loading, filteredTickets} = useAppSelector(state => state.ticketsList)

  useEffect(() => {
    dispatch(fetchTickets());
  }, [])

  if (loading) {
    return <h1>Loading...</h1>
  }

  if (error) {
    return <h1>Error</h1>
  }

  return (
    <div className={styles.list}>
      {filteredTickets
        .map((ticket, i): JSX.Element => (
          <div key={i} className={styles.ticket}>
            <div>
              <div className={styles.ticket__sideContent}>
                <div className={styles.ticket__container}>
                  <div className={styles.ticket__buyButton}>
                    <a href="#" className={styles.ticket__buyButton_btn}>Купить {ticket.price}</a>
                  </div>
                  <div className={styles.ticket__carrier}>Компания: {ticket.carrier}</div>
                  <div className={styles.ticket__stop}>Количество пересдок: {ticket.stops}</div>
                </div>
              </div>
              <div className={styles.ticket__content}>
                <div className={styles.ticket__body}>
                  <div className={styles.ticket__origin}>
                    <div className={styles.ticket__time}>Время отправления: {ticket.departure_time}</div>
                    <div className={styles.ticket__name}>{ticket.origin}, {ticket.origin_name}</div>
                    <div className={styles.ticket__date}>{ticket.departure_date}</div>
                  </div>
                  <div className={styles.ticket__destination}>
                    <div className={styles.ticket__time}>Время прибытия: {ticket.arrival_time}</div>
                    <div className={styles.ticket__name}>{ticket.destination}, {ticket.destination_name}</div>
                    <div className={styles.ticket__date}>{ticket.arrival_date}</div>
                  </div>
                </div>
              </div>
            </div>
          </div>)
      )}
    </div>
  );
};

export default TicketsList;