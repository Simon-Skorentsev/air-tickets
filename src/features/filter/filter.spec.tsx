import ticketsListReduser, { changeCurrence, transferFilter } from '../ticketsList/ticketsListSlice';
import { render } from "@testing-library/react";
import { Provider } from "react-redux";
import TicketsList from "../ticketsList/TicketsList";
import { configureStore } from "@reduxjs/toolkit";
import { Store } from '../../app/store';
import { TicketsState } from '../../types/ticket';

describe("currency and transfers", () => {
    const initialState: TicketsState = {
        tickets: [{
            "origin": "VVO",
            "origin_name": "Владивосток",
            "destination": "TLV",
            "destination_name": "Тель-Авив",
            "departure_date": "12.05.18",
            "departure_time": "16:20",
            "arrival_date": "12.05.18",
            "arrival_time": "22:10",
            "carrier": "TK",
            "stops": 3,
            "price": 12400,
            "initialPrice": 12400
        }],
        filteredTickets: [{
            "origin": "VVO",
            "origin_name": "Владивосток",
            "destination": "TLV",
            "destination_name": "Тель-Авив",
            "departure_date": "12.05.18",
            "departure_time": "16:20",
            "arrival_date": "12.05.18",
            "arrival_time": "22:10",
            "carrier": "TK",
            "stops": 3,
            "price": 12400,
            "initialPrice": 12400
        }],
        loading: false,
        error: null
    };

    const renderWithRedux = (
        component: JSX.Element,
        store: Store
    ) => {
        return {
            ...render(<Provider store={store}>{component}</Provider>)
        }
    }

    it("shoud retain the currency of the ticket after changing the number of transfers of the ticket", () => {
        let actual = ticketsListReduser(initialState, changeCurrence("usd"));
        const dollarPrice = actual.filteredTickets[0].price;

        actual = ticketsListReduser(actual, transferFilter(3));
        expect(actual.filteredTickets[0].price).toEqual(dollarPrice);
    });

    it("shoud save initial price after rub => $ => eur => rub && correct rounding of currency", () => {
        let actual = ticketsListReduser(initialState, changeCurrence("rub"));
        const rubPrice = actual.tickets[0].price;

        actual = ticketsListReduser(initialState, changeCurrence("usd"));
        expect(actual.tickets[0].price.toString().split(".")[1]).toHaveLength(2);  //проверка округления

        actual = ticketsListReduser(actual, changeCurrence("eur"));
        expect(actual.tickets[0].price.toString().split(".")[1]).toHaveLength(2);

        actual = ticketsListReduser(actual, changeCurrence("rub"));
        expect(actual.tickets[0].price === rubPrice);
    });

    it("tickets shouldn't be lost after several changes of transfers", () => {
        let actual = ticketsListReduser(initialState, transferFilter("all"));

        actual = ticketsListReduser(initialState, transferFilter(1));
        expect(actual.filteredTickets.length).toEqual(0);

        actual = ticketsListReduser(initialState, transferFilter(3));
        expect(actual.filteredTickets.length).toEqual(1);
    });
    it("when the number of transfers changes, tickets shouldn't be lost", async () => {
        const store = configureStore({
            reducer: {
                ticketsList: ticketsListReduser
            },
            preloadedState: { ticketsList: initialState }
        });

        const component = renderWithRedux(<TicketsList />, store);  //"Loading..."
        await component.findByText(/13600/);  //render all tickets

        store.dispatch(transferFilter(0));
        store.dispatch(transferFilter(1));
        store.dispatch(transferFilter(2));
        store.dispatch(transferFilter(3));
        store.dispatch(transferFilter(0));

        expect(component.baseElement).toMatchSnapshot();
    });
});
