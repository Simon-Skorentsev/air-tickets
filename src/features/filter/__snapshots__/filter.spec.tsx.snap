// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`currency and transfers when the number of transfers changes, tickets shouldn't be lost 1`] = `
<body>
  <div>
    <div
      class="list"
    >
      <div
        class="ticket"
      >
        <div>
          <div
            class="ticket__sideContent"
          >
            <div
              class="ticket__container"
            >
              <div
                class="ticket__buyButton"
              >
                <a
                  class="ticket__buyButton_btn"
                  href="#"
                >
                  Купить 
                  15300
                </a>
              </div>
              <div
                class="ticket__carrier"
              >
                Компания: 
                SU
              </div>
              <div
                class="ticket__stop"
              >
                Количество пересдок: 
                0
              </div>
            </div>
          </div>
          <div
            class="ticket__content"
          >
            <div
              class="ticket__body"
            >
              <div
                class="ticket__origin"
              >
                <div
                  class="ticket__time"
                >
                  Время отправления: 
                  12:10
                </div>
                <div
                  class="ticket__name"
                >
                  VVO
                  , 
                  Владивосток
                </div>
                <div
                  class="ticket__date"
                >
                  12.05.18
                </div>
              </div>
              <div
                class="ticket__destination"
              >
                <div
                  class="ticket__time"
                >
                  Время прибытия: 
                  18:10
                </div>
                <div
                  class="ticket__name"
                >
                  TLV
                  , 
                  Тель-Авив
                </div>
                <div
                  class="ticket__date"
                >
                  12.05.18
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        class="ticket"
      >
        <div>
          <div
            class="ticket__sideContent"
          >
            <div
              class="ticket__container"
            >
              <div
                class="ticket__buyButton"
              >
                <a
                  class="ticket__buyButton_btn"
                  href="#"
                >
                  Купить 
                  14250
                </a>
              </div>
              <div
                class="ticket__carrier"
              >
                Компания: 
                TK
              </div>
              <div
                class="ticket__stop"
              >
                Количество пересдок: 
                0
              </div>
            </div>
          </div>
          <div
            class="ticket__content"
          >
            <div
              class="ticket__body"
            >
              <div
                class="ticket__origin"
              >
                <div
                  class="ticket__time"
                >
                  Время отправления: 
                  6:10
                </div>
                <div
                  class="ticket__name"
                >
                  VVO
                  , 
                  Владивосток
                </div>
                <div
                  class="ticket__date"
                >
                  12.05.18
                </div>
              </div>
              <div
                class="ticket__destination"
              >
                <div
                  class="ticket__time"
                >
                  Время прибытия: 
                  15:25
                </div>
                <div
                  class="ticket__name"
                >
                  TLV
                  , 
                  Тель-Авив
                </div>
                <div
                  class="ticket__date"
                >
                  12.05.18
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
      <div
        class="ticket"
      >
        <div>
          <div
            class="ticket__sideContent"
          >
            <div
              class="ticket__container"
            >
              <div
                class="ticket__buyButton"
              >
                <a
                  class="ticket__buyButton_btn"
                  href="#"
                >
                  Купить 
                  17400
                </a>
              </div>
              <div
                class="ticket__carrier"
              >
                Компания: 
                S7
              </div>
              <div
                class="ticket__stop"
              >
                Количество пересдок: 
                0
              </div>
            </div>
          </div>
          <div
            class="ticket__content"
          >
            <div
              class="ticket__body"
            >
              <div
                class="ticket__origin"
              >
                <div
                  class="ticket__time"
                >
                  Время отправления: 
                  6:10
                </div>
                <div
                  class="ticket__name"
                >
                  VVO
                  , 
                  Владивосток
                </div>
                <div
                  class="ticket__date"
                >
                  12.05.18
                </div>
              </div>
              <div
                class="ticket__destination"
              >
                <div
                  class="ticket__time"
                >
                  Время прибытия: 
                  16:15
                </div>
                <div
                  class="ticket__name"
                >
                  TLV
                  , 
                  Тель-Авив
                </div>
                <div
                  class="ticket__date"
                >
                  12.05.18
                </div>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</body>
`;
